### ISSUE TITLE
CUDA 7.5 fails with pip install and docker (Ubuntu 14.04)

### ISSUE THREAD

**Issue by soumith** [tensorflow/tensorflow#20]
Installing via:

```bash
# For GPU-enabled version (only install this version if you have the CUDA sdk installed)
$ pip install https://storage.googleapis.com/tensorflow/linux/gpu/tensorflow-0.5.0-cp27-none-linux_x86_64.whl
```

Tried to run the alexnet_benchmark.py and it's looking for CUDA 7.0 specifically.

I have CUDA 7.5 on my machine.

Full stack:
```bash
Traceback (most recent call last):
  File "alexnet_benchmark.py", line 21, in <module>
    import tensorflow.python.platform
  File "/home/awesomebox/anaconda/lib/python2.7/site-packages/tensorflow/__init__.py", line 4, in <module>
    from tensorflow.python import *
  File "/home/awesomebox/anaconda/lib/python2.7/site-packages/tensorflow/python/__init__.py", line 22, in <module>
    from tensorflow.python.client.client_lib import *
  File "/home/awesomebox/anaconda/lib/python2.7/site-packages/tensorflow/python/client/client_lib.py", line 35, in <module>
    from tensorflow.python.client.session import InteractiveSession
  File "/home/awesomebox/anaconda/lib/python2.7/site-packages/tensorflow/python/client/session.py", line 11, in <module>
    from tensorflow.python import pywrap_tensorflow as tf_session
  File "/home/awesomebox/anaconda/lib/python2.7/site-packages/tensorflow/python/pywrap_tensorflow.py", line 28, in <module>
    _pywrap_tensorflow = swig_import_helper()
  File "/home/awesomebox/anaconda/lib/python2.7/site-packages/tensorflow/python/pywrap_tensorflow.py", line 24, in swig_import_helper
    _mod = imp.load_module('_pywrap_tensorflow', fp, pathname, description)
ImportError: libcudart.so.7.0: cannot open shared object file: No such file or directory
```

Tried the docker install, but the docker image is configured for a particular NVIDIA driver version, and doesn't work with others. (this is a known issue: docker driver version and system driver version must exactly match)

**Comment by nivwusquorum** [comment]
Temporary workaround. Go to your library dir (/usr/local/cuda/lib64 or sth like that) and run:
```bash
for x in  *.so.7.5; do sudo ln -s $x ${x::-7}.so.7.0; done
```

**Comment by soumith** [comment]
@nivwusquorum haha that's a terrible workaround, as it'll start issues with other libraries. Thanks a lot though. I'm installing CUDA 7.0

**Comment by lukesimo** [comment]
@nivwusquorum no.

I'm running into the same issue. Looks like I'll be downgrading then.

**Comment by ebrevdo** [comment]
Out of curiosity, have you set your LD_LIBRARY_PATH to your cuda installation's lib64 directory?

**Comment by lukesimo** [comment]
@ebrevdo yes

```
printenv LD_LIBRARY_PATH
/usr/local/cuda-7.5/lib64
```

**Comment by mdda** [comment]
On the subject of CUDA version libraries ...  CUDA 7.0 works for me (as expected), but it really insists on cuDNN 6.5 (which Nvidia now has as 'legacy').  

Exact same library locations, etc, but downgrading from cuDNN 7.0 to 6.5 worked.

**Comment by graphific** [comment]
yes its within the tensorflow code, so just some simple python hacking wont solve it :)
(_pywrap_tensorflow.so when you pip install the binary):

    _mod = imp.load_module('_pywrap_tensorflow', fp, pathname, description)
    ImportError: libcudart.so.7.0: cannot open shared object file: No such file or directory

**Comment by emergix** [comment]
i assume i have same problem: 
I have the 7.5 installed  with tensorflow and when I try (like in the tutorial about gpu)
with tf.device('/gpu:0'):
  a = tf.constant([1.0, 2.0, 3.0, 4.0, 5.0, 6.0], shape=[2, 3], name='a')
  b = tf.constant([1.0, 2.0, 3.0, 4.0, 5.0, 6.0], shape=[3, 2], name='b')
  c = tf.matmul(a, b)
  print(c)
  sess.run(c)

it breaks !
(in torch7, I have no pbs with gpus)

**Comment by jimdsouza** [comment]
Can we assume tensorflow to be forward compatible with cuda 7.5?

**Comment by andorremus** [comment]
I've got the same problem. 

So are you saying that downgrading from cuda 7.5 to 7 should do the trick?

**Comment by androidpcguy** [comment]
I can't seem to find the archive page for the cuda toolkit 7.0? Anyone have a link to it?

**Comment by andorremus** [comment]
If it helps, I've installed cuda toolkit 7.0 and changed the bash profile reference to the new one and it works.

**Comment by ebrevdo** [comment]
Long story short: tensorflow currently requires cuda 7.0.  If you install version 7.0 in a separate directory from 7.5, and point tensorflow at it via the configure script (or LD_LIBRARY_PATH), it will work.  Leaving this open to track future upgrades to the 7.5 SDK.

**Comment by FabHan** [comment]
I'm curious why it is hard to upgrade to CUDA 7.5 and CuDNN v3? Anyone helps me understand?

**Comment by pannous** [comment]
@emergix so you got it to work just by providing libcudart.so.7.0 without reinstalling / downgrading to old cuda?

**Comment by andorremus** [comment]
Yes. It doesn't require you to uninstall the previous one. You can just install them separately and reference v7.0 in your bashrc file

**Comment by pannous** [comment]
Thanks! Workaround confirmed here.

**Comment by fivejjs** [comment]
symbolic link libcuda*.7.5 to libcuda*.7.0
It works.

**Comment by esube** [comment]
This issue and co. are open for more than two months now. Is there any progress to support 7.5 cuda and 7.0 cudnn other than the workarounds? This could be a turn off for some people who already have been working with 7.5 cuda for a while.

**Comment by kmhofmann** [comment]
I completely agree. Ideally (assuming reasonable API/ABI stability on NVIDIA's side), TensorFlow should not be dependent on specific older versions of CUDA and cuDNN. (I'd rather understand it if the *latest* version was required to make use of certain features.)
This is the one issue that puts me off using TensorFlow with GPU support.

**Comment by cmcneil** [comment]
+1 Please support CUDA 7.5

**Comment by ville-k** [comment]
If you want to try out CUDA 7.5 under Linux,  you could try building my pull request branch:
https://github.com/tensorflow/tensorflow/pull/664
It adds support for CUDA on OSX and uses CUDA 7.5 when building under OSX. I haven't tried 7.5 under Linux,  but it seems to work ok with OSX. 
The only change you'd need to make is to edit the configure file and set CUDA_VERSION='7.5' when Linux is detected.  Lines 48-49 would look like:

    if [ "$OSNAME" == "Linux" ]; then
      CUDA_VERSION='7.5'

**Comment by wangg12** [comment]
@ebrevdo @andorremus Could you share the method of installing different cuda(7.0 and 7.5) separately? 

@emergix @fivejjs Does symbolic linking libcuda.7.5 to libcuda.7.0 or just providing libcuda.7.0 have side effects?

**Comment by andorremus** [comment]
@wangg12 A simple google search "install cuda 7.0/7.5 for 'OS here'" should surface all of the information you need

**Comment by wangg12** [comment]
@andorremus I can find a lot of instructions for installing a single version cuda, but I can't find any information about installing 7.0 and 7.5 simultaneously. How do you make it?

**Comment by andorremus** [comment]
Well you install them separately as they are both standalone and the you point to the one you want in the bashrc file afterwards

**Comment by wangg12** [comment]
@andorremus Thanks.

**Comment by wangg12** [comment]
@andorremus Sorry to bother you. One more question, do I need to install driver when I install the second cuda version or I just need to install the libraries?

**Comment by andorremus** [comment]
I think if you've already got the drivers it should be only the libraries
but I'm not really sure on that.

**Comment by esube** [comment]
Just need the toolkit and cudnn. You only need one driver (the latest for
your system, that is)

**Comment by wangg12** [comment]
Thank you @andorremus @esube

**Comment by ducha-aiki** [comment]
@wangg12
could you please clarify, at the end, CUDA 7.0 is needed for only installing, or operate as well? Same for cudnn. My concerns are having two drivers (your and other experiences say it is OK) and performance. CuDNN 2 and 3/4 differs really a lot. Sorry for stupid question :)

@bhack thanks for tagging!

**Comment by ihciah** [comment]
Linking libcudart.so.7.0 to 7.5 solved the problem.
And maybe it should also link libcufft and some other files.
Please support cuda 7.5 and cudnn v3

**Comment by joelclim** [comment]
Installing the download for Ubuntu 14.10 works for me. After installing 7.5 as prescribed on NVIDIA's website. I downloaded the Local Package Installer and used the Synaptic Package Manager to install. My installation looks like this:

lrwxrwxrwx  1 root root    8 Feb  2 15:27 cuda -> cuda-7.0
drwxr-xr-x 13 root root 4096 Feb  2 15:27 cuda-7.0
drwxr-xr-x 13 root root 4096 Feb  2 15:09 cuda-7.5

Make sure the PATH and LD_LIBRARY_PATH variables reflect the right location of the cuda toolkit you use. Using the path the cuda link works very well.

**Comment by vrv** [comment]
@zheng-xq recently added a feature so that:
```
TF_UNOFFICIAL_SETTING=1 ./configure
```

allows you to configure which version of cuda/cudnn/etc to use without requiring manual symlinks.

**Comment by aboulch** [comment]
Hello,

I replaced cuda 7.0 -> 7.5 and cudnn 6.5 -> 7.0 directly in the code.
Seems to work on ubuntu 14.04.
https://github.com/aboulch/tensorflow

**Comment by esafak** [comment]
I pip installed tf 0.7 with CUDA 7.5 and cuDNN4 on Ubuntu 14.04, and tf still sought CUDA 7.0 (libcudart.so.7.0). I had to symlink it to the 7.5 version to get it to work.

**Comment by timsainb** [comment]
@esafak how dod you do that?

**Comment by EderSantana** [comment]
@esafak do you mind explaining what you did? Perhaps in a gist?

**Comment by vrv** [comment]
The pip packages for 0.7.0 should be unversioned -- they should look for libcudart.so, do you have logs to suggest this isn't happening?

**Comment by timsainb** [comment]
```
(tensorflow)tsainbur@txori:~$ pip freeze | grep tensorflow
tensorflow==0.7.0
```

Installed using the virtualenv method, where I had to rename the whl to:
`pip install --upgrade tensorflow-0.7.0-cp27-none-linux_x86_64.whl ` 
to avoid
 `tensorflow-0.7.0-py2-none-linux_x86_64.whl is not a supported wheel on this platform.`

Now I get
 `ImportError: libcudart.so.7.0: cannot open shared object file: No such file or directory`

**Comment by vrv** [comment]
Weird, we were supposed to have built the wheels unversioned.  https://github.com/tensorflow/tensorflow/search?utf8=%E2%9C%93&q=libcudart shows no missing hardcoding of 7.0 anywhere...

@martinwicke, @jendap, any ideas?

**Comment by esafak** [comment]
@timsainb I changed to the appropriate directory and executed `sudo ln -s libcudart.so.7.5 libcudart.so.7.0`. But you might not have to do this if the next paragraph applies.

@EderSantana @vrv I found the problem: the new installation of CUDA did not overwrite or otherwise remove CUDA 7.0, but installed it in /usr/local/cuda7.5 leaving /usr/local/cuda, which got picked up by $LD_LIBRARY_PATH. I removed the old files using

    sudo apt remove cuda-command-line-tools-7-0 cuda-core-7-0 cuda-cublas-7-0 cuda-cublas-dev-7-0 cuda-cudart-7-0 cuda-cudart-dev-7-0 cuda-cufft-7-0 cuda-cufft-dev-7-0 cuda-curand-7-0 cuda-curand-dev-7-0 cuda-cusolver-7-0 cuda-cusolver-dev-7-0 cuda-cusparse-7-0 cuda-cusparse-dev-7-0 cuda-documentation-7-0 cuda-driver-dev-7-0 cuda-license-7-0 cuda-misc-headers-7-0 cuda-npp-7-0 cuda-npp-dev-7-0 cuda-nvrtc-7-0 cuda-nvrtc-dev-7-0 cuda-repo-ubuntu1404-7-0-local cuda-samples-7-0 cuda-toolkit-7-0 cuda-visual-tools-7-0

Now it seems to work.

**Comment by bhack** [comment]
Sede https://github.com/tensorflow/tensorflow/pull/1166

**Comment by tboquet** [comment]
I have the same problem using a docker container and the nvidia-docker tool with an image based on the `cuda:7.5-cudnn4-devel` image. I tried to use @esafak suggestion and I endup with:
```
root@9ad6f042d266:/usr/local/cuda/lib64# readlink -f libcudart.so.7.0
/usr/local/cuda-7.5/targets/x86_64-linux/lib/libcudart.so.7.5.18

root@9ad6f042d266:/usr/local/cuda/lib64# readlink -f libcudart.so
/usr/local/cuda-7.5/targets/x86_64-linux/lib/libcudart.so.7.5.18
```
But i still have:

```python
...
  File "/opt/conda/lib/python2.7/site-packages/tensorflow/python/pywrap_tensorflow.py", line 24, in swig_import_helper
    _mod = imp.load_module('_pywrap_tensorflow', fp, pathname, description)
ImportError: libcudart.so.7.0: cannot open shared object file: No such file or directory
```

If I run python in `/usr/local/cuda/lib64` I have errors related to #808:
```python
I tensorflow/stream_executor/dso_loader.cc:105] successfully opened CUDA library libcublas.so locally
I tensorflow/stream_executor/dso_loader.cc:105] successfully opened CUDA library libcudnn.so locally
I tensorflow/stream_executor/dso_loader.cc:105] successfully opened CUDA library libcufft.so locally
I tensorflow/stream_executor/dso_loader.cc:99] Couldn't open CUDA library libcuda.so. LD_LIBRARY_PATH: /usr/local/nvidia/lib:/usr/local/nvidia/lib64:
I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:121] hostname: 9ad6f042d266
I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:146] libcuda reported version is: Not found: was unable to find libcuda.so DSO loaded into this program
I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:257] driver version file contents: """NVRM version: NVIDIA UNIX x86_64 Kernel Module  352.63  Sat Nov  7 21:25:42 PST 2015
GCC version:  gcc version 4.8.4 (Ubuntu 4.8.4-2ubuntu1~14.04)
"""
I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:150] kernel reported version is: 352.63
I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1051] LD_LIBRARY_PATH: /usr/local/nvidia/lib:/usr/local/nvidia/lib64:
I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1052] failed to find libcuda.so on this system: Failed precondition: could not dlopen DSO: libcuda.so; dlerror: libcuda.so: cannot open shared object file: No such file or directory
I tensorflow/stream_executor/dso_loader.cc:105] successfully opened CUDA library libcurand.so locally
```

Did you manage to find the problem?

Tx!

**Comment by ruffsl** [comment]
@tboquet , your `unable to find libcuda.so` might be related to the bottom of this rabbit hole: https://github.com/tensorflow/tensorflow/issues/970#issuecomment-185862391

**Comment by tboquet** [comment]
@ruffsl, tx, I will try to build the devel version and post my results.

**Comment by vrv** [comment]
Release 0.7.1 wheels now are built assuming cuda-7.5 and cudnn v4.  If you want earlier versions supported, you can build from sources and set the appropriate library paths / versions during ./configure.

**Comment by marekpolewczyk** [comment]
I installed version tensorflow-0.7.1 and CUDA 7.5 and get the message "ImportError: libcudart.so.7.5: cannot open shared object file: No such file or directory". Environment variables are set properly

**Comment by zdx3578** [comment]
For this error
ImportError: libcudart.so.7.5: cannot open shared object file: No such file or directory

I have executed this command to solve it
sudo ldconfig /usr/local/cuda/lib64

**Comment by talpay** [comment]
Shouldn't configure be updated? It's still hardcoded to ask for cuda-7.0 and cudnn-6.5

**Comment by zheng-xq** [comment]
If you build from the latest TF, it would ask you for Cuda SDK and Cudnn versions. You can point it to a different version.

**Comment by monajalal** [comment]
@fivejjs  how did you do so? can you please write the detailed command?

**Comment by brando90** [comment]
why does ldconfig /usr/local/cuda/lib64 solve it @zdx3578

**Comment by philoniare** [comment]
I have just installed CUDA v8 and the issue still persists. I got the following error:
```
ImportError: libcudart.so.7.5: cannot open shared object file: No such file or directory
```
In case anyone else encounters the same issue, the solution is to create symlink. 
1. Go to the cuda directory: `cd /usr/local/cuda/lib64`
2. Create the symlink: `sudo ln -s libcudart.so libcudart.so.7.5`
Then, you'll be able to import tensorflow. To check if it is using the GPU, try running:
```
sess = tf.Session(config=tf.ConfigProto(log_device_placement=True))
```
